### DISCUSSION TITLE
Deploy statically to S3 with prebuild enabled

### DISCUSSION THREAD

**Discussion by snake-py** [vikejs/vike#1007] [Discussion] [answered]
Hello,

Currently we are trying to deploy the app statically to S3 and we have the plugin set up like this. But we are seeing some odd errors, telling us that the hydration is not possible. But locally it works fine after the build, so I guess some cloudfront settings need to be done.

Here is our vite config set up for VPS

```ts
      ssr({
        prerender: true
      })
``` 
We build the app via `tsc && vite build` and then upload the from dist folder the content of the client folder. So inside the S3 it looks like this:

![image](https://github.com/brillout/vite-plugin-ssr/assets/59285387/eecf81ac-0317-4729-b02d-8511cc3e3760)

The bucket allows all access 

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": "*",
            "Action": "s3:GetObject",
            "Resource": "arn:aws:s3:::OUR_BUCKET/*"
        }
    ]
}
```

The S3 content is served in a cloudfront distribution (we know from other vue projects that we need to set up some special error handling pages is anything required here?)

The errors we are seeing on deployed version while locally it works fine. 

```
- Text content does not match server-rendered HTML.
- Hydration failed because the initial UI does not match what was rendered on the server.
- There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root will switch to client rendering
```

Maybe it would be good to have some guide how to deploy this statically to S3?

**Comment by brillout** [comment]
You can check whether your app is running fine locally with `$ vite preview`. If the preview works without errors, then this should also be the case on your static host. (Assuming your satic host doesn't do any HTML rewriting.)

> while locally it works fine

Since `$ vite dev` doesn't do pre-rendering, it's very much expected to have a discrepancy between dev and prod. Using `$ vite preview` is the way to go for testing the prod bulid locally.

Btw. I don't know if I asked you before, but is [sponsoring](https://github.com/sponsors/brillout) something your employer would be up to?

**Reply by snake-py** [reply]
@brillout I actually already planned to ask them, but I thought before asking them I want to have something deployed and ready to properly show case.
 

I will chek if `vite preview` will show me any errors

**Reply by brillout** [reply]
Neat, keep me updated how it goes (also if it's a no).

**Reply by snake-py** [reply]
@brillout so I checked with `vite preview` and it works fine. Just to double check it is possible to deploy the application statically without any server?

**Reply by brillout** [reply]
Yes: https://vite-plugin-ssr.com/pre-rendering.

**Reply by snake-py** [reply]
@brillout Yes I have prerender enabled. Is there then maybe anything special to configure when dealing here with AWS do you maybe know?

**Reply by snake-py** [reply]
@brillout I found my issue it seems related to localization. This SSR thingy is a headache :laughing:

**Reply by brillout** [reply]
👍 Yea, although, the fact that VPS only has low-level i18n APIs isn't helping either. Hopefully (and I believe so) the higher-level i18n API I've in mind will make things easier.

**Reply by snake-py** [reply]
@brillout Do you have an idea how can solve the hydration issue?

**Reply by snake-py** [reply]
This is my set up 

```ts
import i18n from 'i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import Backend from 'i18next-http-backend';
import { initReactI18next } from 'react-i18next';

export const defaultNS = 'client';
export const fallbackNS = 'common';
i18n
  .use(Backend)
  // detect user language
  // learn more: https://github.com/i18next/i18next-browser-languageDetector
  .use(LanguageDetector)
  // pass the i18n instance to react-i18next.
  .use(initReactI18next)
  // init i18next
  // for all options read: https://www.i18next.com/overview/configuration-options
  .init({
    // lng: undefined,
    // backend: {
    //   loadPath: '/locales/{{lng}}/{{ns}}.json'
    // },
    // files to load
    ns: [defaultNS, fallbackNS],
    fallbackLng: ['en'],
    // default namespace (needs no prefix on calling t)
    defaultNS,
    // fallback, can be a string or an array of namespaces
    fallbackNS,
    // logs info level to console output. Helps finding issues with loading not working.
    // debug: true,
    interpolation: {
      escapeValue: false // react already safes from xss => https://www.i18next.com/translation-function/interpolation#unescape
    },
    returnEmptyString: false,
    returnNull: false,
    saveMissing: true
  });

export default i18n;
```

**Reply by snake-py** [reply]
I think the simplest way would be to do it like i18next does it for NextJs - https://github.com/i18next/next-i18next, I guess then VPS would not need a custom API. What do you think?

**Reply by brillout** [reply]
If you pre-render *all* your pages, then the content of every page is static and `$ vite preview` is basically acting like a static host.

So if you don't get hydration mismatches with `$ vite preview`, then you shouldn't get any with your static host either. I'd double check whether your HTML somehow get's transformed. (I would't expect so since you're using S3, but maybe you use another AWS service that proxies & transforms your HTML.)

As for i18n, I'll reply later.

**Reply by brillout** [reply]
Also check https://vite-plugin-ssr.com/hydration-mismatch#common-causes-solutions. Keep me updated so that we can udpate the docs.

**Reply by snake-py** [reply]
@brillout I actually now understand why I get hydration missmatches, it is due to i18n. During the prerender the keys are not being substituted they are inserted and then on first client render they will be updated. Hence I get the hydration error (which I now also see in `vite preview`).

**Reply by snake-py** [reply, marked as answer]
Okay so we had multiple issues here which needed to be fixed. I cannot provide a generalized answer to my issue. 

The two main parts of the issue were:

- Localization was only happening on client side, there fore I implemented it like in the example from the documentation and then it worked
-  the other issue was that some lib component were not ssr ready, therefore we implemented a ClientOnly component like recommended in the docs.
